core/src/main/java/org/jruby/ir/IRMethod.java

```diff
@@ -33,25 +33,28 @@ public IRMethod(IRManager manager, IRScope lexicalParent, DefNode defn, ByteList
 
     @Override
     public boolean hasBeenBuilt() {
-        return defn == null;
+        return defNode == null;
     }
 
-    public synchronized InterpreterContext lazilyAcquireInterpreterContext() {
-        if (!hasBeenBuilt()) {
-            IRBuilder.topIRBuilder(getManager(), this).defineMethodInner(defn, getLexicalParent(), getFlags().contains(IRFlags.CODE_COVERAGE));
-
-            defn = null;
-        }
+    public final InterpreterContext lazilyAcquireInterpreterContext() {
+        if (!hasBeenBuilt()) buildMethodImpl();
 
         return interpreterContext;
     }
 
-    public synchronized BasicBlock[] prepareForCompilation() {
-        if (!hasBeenBuilt()) lazilyAcquireInterpreterContext();
+    private synchronized void buildMethodImpl() {
+        DefNode defNode = this.defNode;
+        if (defNode == null) return;
```

Comment:
You removed use of commont hasBeenBuilt here.  Was it a perf thing?

Reply:
defNode==nul  does it + since we read the field ...
not completely happy about this but it works fine

Reply:
@kares so this was changed because of worry that this.defNode will end up getting nulled if it kept using the helper method?  Since ti is synchronized that is not an issue though is it?  

My reason for these comments is I like to see hasBeenBuilt vs the code used to check it because it explains what that check is for.

Reply:
not really, been just lazily following read volatile var once in a method automatically.
will change it to suit your needs. thanks for the review.